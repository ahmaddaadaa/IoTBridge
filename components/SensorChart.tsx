import React, { useState, useCallback } from "react";
import { Text, View, Dimensions, TouchableOpacity } from "react-native";
import { LineChart } from "react-native-chart-kit";
import * as Animatable from "react-native-animatable";
import useWebSocket, { SensorData } from "../hooks/useWebSocket";
import useNotifications from "../hooks/useNotifications";
import { styles } from "../styles/styles";

const AnimatedLineChart = Animatable.createAnimatableComponent(LineChart);

const SensorChart: React.FC = () => {
  const [selectedSensors, setSelectedSensors] = useState<string[]>([
    "sensor1",
    "sensor2",
    "sensor3",
  ]);
  const notify = useNotifications();

  const checkForAlerts = useCallback(
    (latestData: SensorData) => {
      let message = "";
      if (latestData.sensor1 >= 0.8) {
        message = "Alert: Sensor 1 value hit 0.8 or greater!";
      } else if (latestData.sensor2 >= 0.8) {
        message = "Alert: Sensor 2 value hit 0.8 or greater!";
      } else if (latestData.sensor3 >= 0.8) {
        message = "Alert: Sensor 3 value hit 0.8 or greater!";
      }
      if (message) {
        notify(message);
      }
    },
    [notify],
  );

  const sensorData = useWebSocket(checkForAlerts);

  const toggleSensor = (sensor: string) => {
    setSelectedSensors((prevSelected) =>
      prevSelected.includes(sensor)
        ? prevSelected.filter((s) => s !== sensor)
        : [...prevSelected, sensor],
    );
  };

  const getDataForChart = () => {
    const limitedData = sensorData.slice(-20);
    if (limitedData.length === 0 || selectedSensors.length === 0) {
      return {
        labels: ["0"],
        datasets: [{ data: [0] }],
      };
    }

    const labels = limitedData.map((_, index) => index.toString());
    const datasets = selectedSensors.map((sensor) => ({
      data: limitedData.map((d) => d[sensor as keyof SensorData]),
      color: (opacity = 1) =>
        sensor === "sensor1"
          ? `rgba(0, 150, 255, ${opacity})`
          : sensor === "sensor2"
            ? `rgba(0, 255, 0, ${opacity})`
            : `rgba(255, 0, 0, ${opacity})`,
      strokeWidth: 2,
    }));

    return { labels, datasets };
  };

  return (
    <View style={styles.chartContainer}>
      <Animatable.View
        animation="fadeIn"
        duration={1000}
        style={styles.chartContainer}
      >
        <AnimatedLineChart
          data={getDataForChart()}
          width={Dimensions.get("window").width - 32}
          height={220}
          yAxisLabel=""
          yAxisSuffix=""
          chartConfig={{
            backgroundColor: "#e0e0e0",
            backgroundGradientFrom: "#e0e0e0",
            backgroundGradientTo: "#ffffff",
            decimalPlaces: 2,
            color: (opacity = 1) => `rgba(0, 150, 255, ${opacity})`,
            labelColor: (opacity = 1) => `rgba(0, 0, 0, ${opacity})`,
            style: {
              borderRadius: 16,
            },
            propsForDots: {
              r: "6",
              strokeWidth: "2",
              stroke: "#0A84FF",
            },
          }}
          bezier
          style={{
            marginVertical: 8,
            borderRadius: 16,
          }}
        />
      </Animatable.View>
      <View style={styles.buttonContainer}>
        {["sensor1", "sensor2", "sensor3"].map((sensor) => (
          <TouchableOpacity key={sensor} onPress={() => toggleSensor(sensor)}>
            <Animatable.View
              animation={selectedSensors.includes(sensor) ? "pulse" : undefined}
              duration={2000}
              iterationCount="infinite"
              style={[
                styles.button,
                selectedSensors.includes(sensor)
                  ? styles.buttonActive
                  : styles.buttonInactive,
              ]}
            >
              <Text style={styles.buttonText}>{sensor.toUpperCase()}</Text>
            </Animatable.View>
          </TouchableOpacity>
        ))}
        <TouchableOpacity
          style={[styles.button, styles.allSensorsButton]}
          onPress={() => setSelectedSensors(["sensor1", "sensor2", "sensor3"])}
        >
          <Text style={styles.buttonText}>All Sensors</Text>
        </TouchableOpacity>
      </View>
    </View>
  );
};

export default SensorChart;
