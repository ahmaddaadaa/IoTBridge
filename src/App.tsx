import React from "react";
import { StyleSheet, Text, View, Alert } from "react-native";
import Button from "./components/Button";

export default function App() {
  const handlePress = () => {
    Alert.alert("Button pressed!");
  };

  return (
    <View style={styles.container}>
      <Text>Open up App.jsx to start working on your app!</Text>
      <Button title="Press Me" onPress={handlePress} />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: "#fff",
    alignItems: "center",
    justifyContent: "center",
  },
});
