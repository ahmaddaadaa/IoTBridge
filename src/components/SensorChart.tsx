import React, { useEffect, useState, useRef, useCallback } from "react";
import { Text, View, Dimensions, TouchableOpacity } from "react-native";
import { LineChart } from "react-native-chart-kit";
import Animated, {
  useSharedValue,
  useAnimatedStyle,
  withTiming,
  Easing,
} from "react-native-reanimated";
import {
  registerForPushNotificationsAsync,
  sendPushNotification,
} from "../../notificationService";
import useGlowStyle from "../../hooks/useGlowStyle";
import { styles } from "../../styles/styles";

type SensorData = {
  timestamp: number;
  sensor1: number;
  sensor2: number;
  sensor3: number;
};

const AnimatedLineChart = Animated.createAnimatedComponent(LineChart);

const SensorChart: React.FC = () => {
  const [sensorData, setSensorData] = useState<SensorData[]>([
    { timestamp: Date.now(), sensor1: 0, sensor2: 0, sensor3: 0 },
  ]);
  const [selectedSensors, setSelectedSensors] = useState<string[]>([
    "sensor1",
    "sensor2",
    "sensor3",
  ]);
  const [expoPushToken, setExpoPushToken] = useState<string | undefined>(
    undefined
  );
  const ws = useRef<WebSocket | null>(null);
  const opacity = useSharedValue(0);

  const checkForAlerts = useCallback(
    (data: SensorData[]) => {
      const latestData = data[data.length - 1];
      let message = "";
      if (latestData.sensor1 >= 0.8) {
        message = "Alert: Sensor 1 value hit 0.8 or greater!";
      } else if (latestData.sensor2 >= 0.8) {
        message = "Alert: Sensor 2 value hit 0.8 or greater!";
      } else if (latestData.sensor3 >= 0.8) {
        message = "Alert: Sensor 3 value hit 0.8 or greater!";
      }

      if (message && expoPushToken) {
        sendPushNotification(expoPushToken, message);
      }
    },
    [expoPushToken]
  );

  useEffect(() => {
    const setupNotifications = async () => {
      try {
        const token = await registerForPushNotificationsAsync();
        setExpoPushToken(token);
      } catch (error) {
        console.error("Error registering for push notifications:", error);
      }
    };

    setupNotifications();

    // Replace 'YOUR_LOCAL_IP' with your actual local IP address
    ws.current = new WebSocket("ws://192.168.4.87:8000");

    ws.current.onopen = () => {
      console.log("WebSocket connection opened");
    };

    ws.current.onmessage = (event) => {
      try {
        const data: SensorData = JSON.parse(event.data);
        if (isValidData(data)) {
          setSensorData((prevData) => {
            const newData = [...prevData.slice(-19), data];
            checkForAlerts(newData);
            return newData;
          });
        } else {
          console.log("Received invalid data:", data);
        }
      } catch (error) {
        console.error("Error parsing WebSocket message:", error);
      }
    };

    ws.current.onerror = (error) => {
      console.error("WebSocket error:", error);
    };

    ws.current.onclose = () => {
      console.log("WebSocket connection closed");
    };

    return () => {
      if (ws.current) {
        ws.current.close();
      }
    };
  }, [checkForAlerts]);

  useEffect(() => {
    opacity.value = withTiming(1, {
      duration: 1000,
      easing: Easing.out(Easing.exp),
    });
  }, [sensorData, opacity]);

  const isValidData = (data: SensorData) => {
    return (
      isFinite(data.sensor1) && isFinite(data.sensor2) && isFinite(data.sensor3)
    );
  };

  const toggleSensor = (sensor: string) => {
    setSelectedSensors((prevSelected) =>
      prevSelected.includes(sensor)
        ? prevSelected.filter((s) => s !== sensor)
        : [...prevSelected, sensor]
    );
  };

  const getDataForChart = () => {
    if (sensorData.length === 0 || selectedSensors.length === 0) {
      return {
        labels: ["0"],
        datasets: [{ data: [0] }], // Provide default values to prevent rendering issues
      };
    }

    const labels = sensorData.map((_, index) => index.toString());
    const datasets = selectedSensors.map((sensor) => ({
      data: sensorData.map((d) => {
        const value = d[sensor as keyof SensorData];
        return isFinite(value) ? value : 0; // Ensure the value is finite
      }),
      color: (opacity = 1) =>
        sensor === "sensor1"
          ? `rgba(0, 150, 255, ${opacity})`
          : sensor === "sensor2"
          ? `rgba(0, 255, 0, ${opacity})`
          : `rgba(255, 0, 0, ${opacity})`,
      strokeWidth: 2,
    }));

    return { labels, datasets };
  };

  const animatedStyle = useAnimatedStyle(() => {
    return {
      opacity: opacity.value,
    };
  });

  const sensor1GlowStyle = useGlowStyle(selectedSensors.includes("sensor1"));
  const sensor2GlowStyle = useGlowStyle(selectedSensors.includes("sensor2"));
  const sensor3GlowStyle = useGlowStyle(selectedSensors.includes("sensor3"));
  const allSensorsGlowStyle = useGlowStyle(selectedSensors.length === 3);

  return (
    <View style={styles.container}>
      <Text style={styles.header}>Real-Time Sensor Data</Text>
      <Animated.View style={[styles.chartContainer, animatedStyle]}>
        <AnimatedLineChart
          data={getDataForChart()}
          width={Dimensions.get("window").width - 32}
          height={220}
          yAxisLabel=""
          yAxisSuffix=""
          chartConfig={{
            backgroundColor: "#e0e0e0",
            backgroundGradientFrom: "#e0e0e0",
            backgroundGradientTo: "#ffffff",
            decimalPlaces: 2,
            color: (opacity = 1) => `rgba(0, 150, 255, ${opacity})`,
            labelColor: (opacity = 1) => `rgba(0, 0, 0, ${opacity})`,
            style: {
              borderRadius: 16,
            },
            propsForDots: {
              r: "6",
              strokeWidth: "2",
              stroke: "#0A84FF",
            },
          }}
          bezier
          style={{
            marginVertical: 8,
            borderRadius: 16,
          }}
        />
      </Animated.View>
      <View style={styles.buttonContainer}>
        <TouchableOpacity onPress={() => toggleSensor("sensor1")}>
          <Animated.View
            style={[
              styles.button,
              selectedSensors.includes("sensor1")
                ? styles.buttonActive
                : styles.buttonInactive,
              sensor1GlowStyle,
            ]}
          >
            <Text style={styles.buttonText}>Sensor 1</Text>
          </Animated.View>
        </TouchableOpacity>
        <TouchableOpacity onPress={() => toggleSensor("sensor2")}>
          <Animated.View
            style={[
              styles.button,
              selectedSensors.includes("sensor2")
                ? styles.buttonActive
                : styles.buttonInactive,
              sensor2GlowStyle,
            ]}
          >
            <Text style={styles.buttonText}>Sensor 2</Text>
          </Animated.View>
        </TouchableOpacity>
        <TouchableOpacity onPress={() => toggleSensor("sensor3")}>
          <Animated.View
            style={[
              styles.button,
              selectedSensors.includes("sensor3")
                ? styles.buttonActive
                : styles.buttonInactive,
              sensor3GlowStyle,
            ]}
          >
            <Text style={styles.buttonText}>Sensor 3</Text>
          </Animated.View>
        </TouchableOpacity>
        <TouchableOpacity
          onPress={() => setSelectedSensors(["sensor1", "sensor2", "sensor3"])}
        >
          <Animated.View
            style={[
              styles.button,
              selectedSensors.length === 3
                ? styles.buttonActive
                : styles.buttonInactive,
              allSensorsGlowStyle,
            ]}
          >
            <Text style={styles.buttonText}>All Sensors</Text>
          </Animated.View>
        </TouchableOpacity>
      </View>
    </View>
  );
};

export default SensorChart;