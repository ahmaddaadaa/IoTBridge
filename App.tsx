import React from "react";
import { StyleSheet, View } from "react-native";
import SensorChart from "./src/components/SensorChart";

const App: React.FC = () => {
  return (
    <View style={styles.container}>
      <SensorChart />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center",
    backgroundColor: "#fff",
  },
});

export default App;
