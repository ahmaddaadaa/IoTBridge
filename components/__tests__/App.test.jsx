/* eslint-env jest */
import React from "react";
import renderer from "react-test-renderer";
import App from "../../App";
import SensorChart from "../../components/SensorChart";

// Mock the alert function
global.alert = (message) => console.log(message);

// Mock the SensorChart component to avoid WebSocket and other side effects
jest.mock("../../components/SensorChart", () => {
  return function DummySensorChart() {
    return <div>SensorChart</div>;
  };
});

describe("<App />", () => {
  it("renders correctly", () => {
    const tree = renderer.create(<App />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("contains the SensorChart component", () => {
    const tree = renderer.create(<App />).root;
    expect(tree.findByType(SensorChart)).toBeTruthy();
  });
});
