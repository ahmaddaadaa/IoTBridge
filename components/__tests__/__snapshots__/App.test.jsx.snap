// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <div>
    SensorChart
  </div>
</View>
`;
